### PR TITLE
FED-1863 Required props lint

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Dependent PR

This PR is dependent on https://github.com/Workiva/over_react/pull/870

---

## Motivation
To help provide a good developer experience when working with required late props, we want to provide a lint that warns when a component is missing required props.

That way, devs can get static warnings, and not wait until they hit our runtime errors to find out they were missing props.

This lint should be similar to the warnings you get when you are missing required function parameters.

We had a partial implementation of this lint, and we needed to get it production-ready.

## Changes
- Finish `MissingRequiredPropDiagnostic` implementation, enable it, add tests
    - This supports both `late` required and `@requiredProp` required props, but after some back and forth I opted to only lint for `late` required by default. For a more detailed explanation, see comment in tools/analyzer_plugin/lib/src/plugin.dart
- Add utilities used by the diagnostic, with unit tests
  - `getAllProps` to get all props for a props class and its supertypes
  - `getAllRequiredProps` to collect info on which props are required
  - `isRequiredPropValidationDisabled`
- Raise analyzer plugin SDK language constraint to 2.17 to use enhanced enums
- Other changes not directly related to this diagnostic
    - Work around InconsistentAnalysisExceptions when typing
        - I noticed when manually testing that we'd get a bunch errors in the IDE about `InconsistentAnalysisException`, unrelated to this PR. I found similar issues documented in Dart SDK tooling, and applied a similar workaround.
    - Improve `debug: over_react_metrics` output to aid in performance optimizing
      - Star the top 3 slowest diagnostics
      - Show `getAllComponentUsages` overhead
    - `AnalyzerDebugHelper` - make log message and location computation lazy by default, so that we're not slowing down performance when debug flags aren't enabled
      - Update existing usages, add locations in missing_cascade_parens.dart
    
    

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        - Manually verify that diagnostic behaves as expected in tools/analyzer_plugin/playground/web/missing_required_props.dart
        - Remove the `over_react_annotation_required_prop` line in the playground's analysis_options.yaml and verify that annotation-required props don't lint by default
        - Try typing in the playground and verify there are virtually no `InconsistentAnalysisExceptions` errors compared to the base branch
        - Verify behavior of debug flag comments looks good:
            - `// debug: over_react_metrics` 
                - Verify new behavior mentioned in Changes section looks good
                - Verify the performance of this diagnostic looks good
            - `// debug: over_react_required_props` - shows information about required props for each builder
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: tools/analyzer_plugin/lib/src/async_plugin_apis/diagnostic.dart
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/async_plugin_apis/diagnostic.dart:1]: I'd recommend using [whitespace-agnostic diff](https://github.com/Workiva/over_react/pull/871/files?w=1#diff-7a8e28d75e77fe3c8eda22871a60c2536a21744c2746b7ec68644913d62eeff1) for this file

File: tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart:1]: Since this diagnostic always existed but was only partially implemented and was always disabled, this full file should be reviewed as if it were newly-added, as opposed to the diff being reviewed

File: tools/analyzer_plugin/lib/src/util/fix.dart
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/util/fix.dart:10]: This typing seemed to be a typo; calls to `buildFileEdit` aren't awaited, and all the existing `buildFileEdit`s were synchronous. This signature now matches that of `buildGenericFileEdit`

File: tools/analyzer_plugin/lib/src/util/prop_declarations/get_all_props.dart
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/util/prop_declarations/get_all_props.dart:4]: I spent quite a bit of time optimizing this function, and I feel pretty happy about its performance! 

In practice, usage is optimized further via the caching done with `_cachedGetAllRequiredProps`.

File: tools/analyzer_plugin/lib/src/async_plugin_apis/diagnostic.dart
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/async_plugin_apis/diagnostic.dart:160]: I'd recommend using [whitespace-agnostic diff](https://github.com/Workiva/over_react/pull/871/files?w=1#diff-7a8e28d75e77fe3c8eda22871a60c2536a21744c2746b7ec68644913d62eeff1R160) for these diffs

File: tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart:56]: Since this diagnostic always existed but was only partially implemented, was always disabled, and was never fully reviewed, this full file should be reviewed as if it were newly-added, as opposed to reviewing just the diff

File: tools/analyzer_plugin/lib/src/util/prop_declarations/required_props.dart
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/lib/src/util/prop_declarations/required_props.dart:82]: This is currently unused, but will get used as part of updates to this file in FED-2015; since I was adding `requirednessLevel` and new `PropRequiredness` enum test cases, I figured I'd just stub it in now.

File: tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart
Comment (aaronlademann-wf) [inline tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart:99]: Does this need to be addressed in this PR?

File: tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart
Comment (aaronlademann-wf) [inline tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart:139]: 💯

File: tools/analyzer_plugin/playground/analysis_options.yaml
Comment (aaronlademann-wf) [inline tools/analyzer_plugin/playground/analysis_options.yaml:10]: Is `info` the default severity of this?

**Review by aaronlademann-wf — APPROVED** [review]
Couple questions.

Insane test coverage. 🥇

File: tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart
Comment (greglittlefield-wf) [reply tools/analyzer_plugin/lib/src/diagnostic/missing_required_prop.dart:99]: No, but I would love to circle back to it later on, maybe next time we add a new debug flag

File: tools/analyzer_plugin/playground/analysis_options.yaml
Comment (greglittlefield-wf) [reply tools/analyzer_plugin/playground/analysis_options.yaml:10]: It doesn't really have a default severity since it's disabled by default, and to enable it you also have to specify a severity

File: tools/analyzer_plugin/playground/web/missing_required_props.dart
Comment (sydneyjodon-wk) [inline tools/analyzer_plugin/playground/web/missing_required_props.dart:53]: I haven't looked at this PR, but I was just messing around in the playground and wondering if it shouldn't be linting when required prop validation is disabled on the usage like this: 
```
(WithLateRequired()..disableRequiredPropValidation())();
```

Comment (greglittlefield-wf) [reply tools/analyzer_plugin/playground/web/missing_required_props.dart:55]: Done in [a7e3f7670f33bb530a738856f01faa23c3fdf2c5](https://github.com/Workiva/over_react/pull/871/commits/a7e3f7670f33bb530a738856f01faa23c3fdf2c5)!

File: tools/analyzer_plugin/test/unit/util/prop_declaration/shared_util.dart
Comment (greglittlefield-wf) [inline tools/analyzer_plugin/test/unit/util/prop_declaration/shared_util.dart:41]: This was cherry-picked from this PR here: https://github.com/Workiva/over_react/pull/873/commits/c859acbcce79d90594cfec8595787536e533e69b to get CI passing after merging the WIP branch, and will get addressed in that PR (FIXME removed in a later commit here https://github.com/Workiva/over_react/pull/873/commits/008939cfcc3b8314ddd6b8e36f29d218ba2e67e8)

**Review by aaronlademann-wf — APPROVED** [review]

**Comment by sydneyjodon-wk** [comment]
Starting QA on this :loading:

**Review by sydneyjodon-wk — APPROVED** [review]
QA +1 Smoke tested 

- [x] Manually verify that diagnostic behaves as expected in tools/analyzer_plugin/playground/web/missing_required_props.dart
- [x] Remove the `over_react_annotation_required_prop` line in the playground's analysis_options.yaml and verify that annotation-required props don't lint by default
- [x] Try typing in the playground and verify there are virtually no `InconsistentAnalysisExceptions` errors compared to the base branch
- [x] Verify behavior of debug flag comments looks good:
  - [x] `// debug: over_react_metrics`
    - [x] Verify new behavior mentioned in Changes section looks good
    - [x] Verify the performance of this diagnostic looks good
  - [x] `// debug: over_react_required_props` - shows information about required props for each builder

**Pull request merged**
